src/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -268,6 +268,9 @@ public SamReader open(final SamInputResource resource) {
                             bufferedStream.close();
                             bufferedStream = null;
                         }
+
+                        bufferedStream = IOUtil.maybeBufferInputStream(indexMaybe.asUnbufferedInputStream());
```

Comment:
You should always close the previous `bufferedStream` connected to `sourceFile` here before creating your new input stream on the index.